bee-api/bee-rest-api/src/endpoints/permission.rs

```diff
@@ -1,30 +1,132 @@
+                // Check if the requested endpoint is open for public use.
+                if args.rest_api_config.public_routes.is_match(path.as_str()) {
+                    return Ok(());
+                }
+
+                // Else the requested endpoint is protected by JWT / or not even exposed.
+                let jwt = extract_jwt(&headers)?;
+                // Decode the JWT payload to find out how to validate it. The `aud` claim will indicate if it's an API
+                // JWT or a Dashboard JWT.
+                let jwt_payload = {
+                    let jwt_string = jwt.to_string();
+                    // Every JWT consists of 3 parts: 1) header, 2) payload, 3) signature.
+                    // The different parts are separated by `.`.
+                    let split = jwt_string.split('.').collect::<Vec<&str>>();
+                    // If there are less or more then 3 parts the given JWT is invalid.
+                    if split.len() != 3 {
+                        return Err(reject::custom(CustomRejection::Forbidden));
+                    }
+
+                    // Get the base64 encoded payload.
+                    let encoded_payload = split[1];
+                    // Base64-decode the payload to bytes.
+                    let decoded_payload_bytes =
+                        base64::decode(encoded_payload).map_err(|_| reject::custom(CustomRejection::Forbidden))?;
+                    // Create a UTF-8 string from the decoded bytes.
+                    String::from_utf8(decoded_payload_bytes).map_err(|_| CustomRejection::Forbidden)?
+                };
+
+                if jwt_payload.contains(&format!("\"aud\":\"{}\"", API_AUDIENCE_CLAIM)) {
+                    if validate_api_jwt(&jwt, &args).is_ok()
+                        && args.rest_api_config.protected_routes.is_match(path.as_str())
+                    {
+                        return Ok(());
+                    }
+                } else {
+                    #[cfg(feature = "dashboard")]
+                    if jwt_payload.contains(&format!("\"aud\":\"{}\"", DASHBOARD_AUDIENCE_CLAIM))
+                        && validate_dashboard_jwt(&jwt, &args).is_ok()
+                        && DASHBOARD_ROUTES.is_match(path.as_str())
+                    {
                         return Ok(());
                     }
                 }
+
                 Err(reject::custom(CustomRejection::Forbidden))
             }
         })
         .untuple_one()
 }
+
+fn extract_jwt(headers: &HeaderMap) -> Result<JsonWebToken, Rejection> {
+    let header = headers.get(AUTHORIZATION).ok_or(CustomRejection::Forbidden)?;
+    let auth_header = std::str::from_utf8(header.as_bytes()).map_err(|_| CustomRejection::Forbidden)?;
+
+    if !auth_header.starts_with(BEARER) {
+        return Err(reject::custom(CustomRejection::Forbidden));
+    }
+
+    Ok(JsonWebToken::from(auth_header.trim_start_matches(BEARER).to_owned()))
```

Comment:
out of interest: BEARER contains a whitespace at the end, right? `trim_start_matches` removes is used to remove it here everytime it's called? I guess there's also a `start_matches`?

Reply:
Hmm seems like there is only `trim_start_matches` https://doc.rust-lang.org/std/primitive.str.html#method.trim_start_matches

Reply:
alright, I was just curious. Ignore me!